web/src/app/components/tasks-editor/tasks-editor.component.ts

```diff
@@ -0,0 +1,241 @@
+/**
+ * Copyright 2023 The Ground Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the 'License');
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an 'AS IS' BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {CdkDragDrop} from '@angular/cdk/drag-drop';
+import {Component, EventEmitter, Input, Output} from '@angular/core';
+import {
+  AbstractControl,
+  FormArray,
+  FormBuilder,
+  FormGroup,
+  Validators,
+} from '@angular/forms';
+import {
+  Cardinality,
+  MultipleChoice,
+} from 'app/models/task/multiple-choice.model';
+import {Option} from 'app/models/task/option.model';
+import {Task, TaskType} from 'app/models/task/task.model';
+import {DataStoreService} from 'app/services/data-store/data-store.service';
+import {DialogService} from 'app/services/dialog/dialog.service';
+import {moveItemInFormArray} from 'app/utils/utils';
+import {List} from 'immutable';
+
+export enum TaskGroup {
+  QUESTION = 1,
+  PHOTO = 2,
+  DROP_PIN = 3,
+  DRAW_AREA = 4,
+  CAPTURE_LOCATION = 5,
+}
+
+export const taskGroupToTypes = new Map([
+  [
+    TaskGroup.QUESTION,
+    List([TaskType.TEXT, TaskType.DATE, TaskType.MULTIPLE_CHOICE]),
+  ],
+  [TaskGroup.PHOTO, List([TaskType.PHOTO])],
+  [TaskGroup.DROP_PIN, List([TaskType.DROP_PIN])],
+  [TaskGroup.DRAW_AREA, List([TaskType.DRAW_AREA])],
+  [TaskGroup.CAPTURE_LOCATION, List([TaskType.CAPTURE_LOCATION])],
+]);
+
+export const taskTypeToGroup = new Map([
+  [TaskType.TEXT, TaskGroup.QUESTION],
+  [TaskType.DATE, TaskGroup.QUESTION],
+  [TaskType.MULTIPLE_CHOICE, TaskGroup.QUESTION],
+  [TaskType.PHOTO, TaskGroup.PHOTO],
+  [TaskType.DROP_PIN, TaskGroup.DROP_PIN],
+  [TaskType.DRAW_AREA, TaskGroup.DRAW_AREA],
+  [TaskType.CAPTURE_LOCATION, TaskGroup.CAPTURE_LOCATION],
+]);
+
+@Component({
+  selector: 'tasks-editor',
+  templateUrl: './tasks-editor.component.html',
+  styleUrls: ['./tasks-editor.component.scss'],
+})
+export class TasksEditorComponent {
+  formGroup!: FormGroup;
+
+  @Input() label?: string;
+  @Input() tasks?: List<Task>;
+  @Output() onChange?: EventEmitter<boolean> = new EventEmitter<boolean>();
+
+  addableTaskGroups: Array<TaskGroup> = [
+    TaskGroup.QUESTION,
+    TaskGroup.PHOTO,
+    TaskGroup.DROP_PIN,
+    TaskGroup.DRAW_AREA,
+    TaskGroup.CAPTURE_LOCATION,
+  ];
+
+  constructor(
+    private dataStoreService: DataStoreService,
+    private dialogService: DialogService
+  ) {}
+
+  ngOnChanges(): void {
+    const formBuilder = new FormBuilder();
+
+    if (this.tasks) {
+      this.formGroup = formBuilder.group({
```

Comment:
This block looks intimidating :) Can we refactor it a bit into local vars and handler functions?